### ISSUE TITLE
Country reappears after it gets back into the viewport

### ISSUE THREAD

**Issue by mindaugasveblauskas** [Mapsui/Mapsui#2714]
**Mapsui Platform**
MAUI, Blazer (Web samples)
https://mapsui.com/samples/

Category: **Styles**
Sample: **ThemeStyle on shapefile**

**Describe the bug**
Countries are hidden when they're not in the viewport anymore and showed again once in the viewport. I assume this is done specifically to optimize performance of the map, but maybe it's possible to change this behavior?

**To Reproduce**
It's easy  to reproduce just by panning the map.

**Expected behavior**
Countries to be kept "visible" when out of viewport.

**Video**
https://github.com/user-attachments/assets/546b0e58-19f8-462e-9370-a572d6e29795

**Comment by pauldendulk** [comment]
You are using the 'Layer' class?

I have seen this too and it looks weird indeed. Perhaps it should be fixed by only removing items from the Layer cache (I think that is where te problem is) if it far outside the current extent. Maybe we should filter on an extent that is grown by the size of the extent, so, it would be nine times bigger.

**Comment by mindaugasveblauskas** [comment]
Yes, I use Layer class. Could you please point me to the "right" place in the source code so I could try to debug and potentially fix this?

**Comment by pauldendulk** [comment]
The Layer fetches the data from the IProvider and puts it in a local cache, replacing the previous cache. If you do not have a large amount of data in the source all data could stay in the Layer once it is fetched. 

I think your best option is to create your own Layer and solve it in a way that suits your data. Perhaps doing just one initial fetch for the whole extent, and using that cache in all later fetches.

**Comment by charlenni** [comment]
Sorry for jumping in, but I think it isn't the responsibility of the Layer to cache data. From my understanding the Layer holds a Provider and Styles for drawing the Features returned by the Provider. 

Btw. that in mind, the MemoryLayer should be removed, because it combines Layer and Provider. Same to WritableLayer.

**Comment by pauldendulk** [comment]
There are a lot of options I am considering to improve this. One thing I have in mind it let the Layer be not much more than a configuration class, and do the fetching and possible caching on the outside. But there are many options and many aspects to keep in mind.

**Comment by mindaugasveblauskas** [comment]
For testing purposes I've implemented this custom GeoJsonProvider and it does prove the idea of increasing the extent actually solves the issue:

```
public class CustomGeoJsonProvider : IProvider
{
    private readonly GeoJsonProvider _geoJsonProvider;

    public CustomGeoJsonProvider(GeoJsonProvider geoJsonProvider)
    {
        _geoJsonProvider = geoJsonProvider;
    }

    public MRect GetExtent()
    {
        return _geoJsonProvider.GetExtent();
    }

    public Task<IEnumerable<IFeature>> GetFeaturesAsync(FetchInfo fetchInfo)
    {
        MSection section = new(fetchInfo.Section.Extent.Grow(200), fetchInfo.Resolution);
        FetchInfo fetchInfoCopy = new FetchInfo(section, fetchInfo.CRS, fetchInfo.ChangeType);

        return _geoJsonProvider.GetFeaturesAsync(fetchInfoCopy);
    }

    public string CRS
    {
        get => _geoJsonProvider.CRS;
        set => _geoJsonProvider.CRS = value;
    }
}
```

However, I would trust the author more on a proper way of fixing this 😄

**Comment by pauldendulk** [comment]
Now that we have centralized data fetching this is easier to implement.
